codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/transformers/RefactorConstrainedMemberType.kt

```diff
@@ -0,0 +1,195 @@
+        val structName = memberShape.name
+        val memberName = memberShape.member.orElse(null)
+            .replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }
+
+        fun makeStructName(suffix: String = "") =
+            ShapeId.from("${memberShape.namespace}#Refactored${structName}${memberName}$suffix")
+
+        fun structNameIsUnique(newName: ShapeId) =
+            model.getShape(newName).isEmpty && !existingRefactoredNames.contains(newName)
+
+        fun generateUniqueName(): ShapeId {
+            // Ensure the name does not already exist in the model, else make it unique
+            // by appending a new number as the suffix.
+            var suffix = ""
+            (0..100).forEach {
+                val extractedStructName = makeStructName(suffix)
+                if (structNameIsUnique(extractedStructName))
+                    return extractedStructName
+
+                suffix = "_$it"
+            }
+
+            // A unique type could not be found by adding a numeric suffix to it. As a
+            // secondary algorithm, use existing struct names as suffix to generate a unique name.
+            suffix = "_"
+
+            model.listShapes
+                .filterIsInstance<StructureShape>()
+                .map {
+                    suffix = suffix.plus(it.id.name)
+
+                    val extractedStructName = makeStructName(suffix)
+                    if (structNameIsUnique(extractedStructName))
+                        return extractedStructName
+                }
+
+            throw IllegalStateException("A unique name for the refactored structure type could not be generated")
+        }
+
+        val newName = generateUniqueName()
+        existingRefactoredNames.add(newName)
+        return newName
+    }
+
+    /**
+     * Returns the transformation that would be required to turn the given member shape
+     * into a non-constrained member shape.
+     */
+    private fun MemberShape.makeNonConstrained(
+        model: Model,
+        existingRefactoredNames: MutableSet<ShapeId>,
+    ): MemberTransformation? {
+        val (constrainedTraits, otherTraits) = this.allTraits.values
+            .partition {
+                allConstraintTraits.contains(it.javaClass)
+            }
+
+        // No transformation required in case the given MemberShape has no constraints.
+        if (constrainedTraits.isEmpty())
+            return null
+
+        // Create a new standalone shape of the same target as the target of the member e.g.
+        // someX : Centigrade, where integer : Centigrade, then the new shape should be
+        // integer: newShape
+        val targetShape = model.expectShape(this.target)
+
+        // The shape has to be a buildable type otherwise it cannot be refactored.
+        if (targetShape is ToSmithyBuilder<*>) {
+            when (val builder = targetShape.toBuilder()) {
+                is AbstractShapeBuilder<*, *> -> {
```

Comment:
This is needed because `.allTraits` and `.id` are only defined on `AbstractShapeBuilder` right?